sw/device/lib/crypto/impl/ecc/ecdsa_p384_verify.c

```diff
@@ -0,0 +1,189 @@
+    OTBN_ADDR_T_INIT(p384_ecdsa_verify, dptr_y);
+static const otbn_addr_t kOtbnVarEcdsaDptrRnd =
+    OTBN_ADDR_T_INIT(p384_ecdsa_verify, dptr_rnd);
+static const otbn_addr_t kOtbnVarEcdsaDptrMsg =
+    OTBN_ADDR_T_INIT(p384_ecdsa_verify, dptr_msg);
+static const otbn_addr_t kOtbnVarEcdsaDptrR =
+    OTBN_ADDR_T_INIT(p384_ecdsa_verify, dptr_r);
+static const otbn_addr_t kOtbnVarEcdsaDptrS =
+    OTBN_ADDR_T_INIT(p384_ecdsa_verify, dptr_s);
+static const otbn_addr_t kOtbnVarEcdsaX =
+    OTBN_ADDR_T_INIT(p384_ecdsa_verify, x);
+static const otbn_addr_t kOtbnVarEcdsaY =
+    OTBN_ADDR_T_INIT(p384_ecdsa_verify, y);
+static const otbn_addr_t kOtbnVarEcdsaMsg =
+    OTBN_ADDR_T_INIT(p384_ecdsa_verify, msg);
+static const otbn_addr_t kOtbnVarEcdsaR =
+    OTBN_ADDR_T_INIT(p384_ecdsa_verify, r);
+static const otbn_addr_t kOtbnVarEcdsaS =
+    OTBN_ADDR_T_INIT(p384_ecdsa_verify, s);
+static const otbn_addr_t kOtbnVarEcdsaRnd =
+    OTBN_ADDR_T_INIT(p384_ecdsa_verify, rnd);
+
+enum {
+  /*
+   * Mode is represented by a single word.
+   */
+  kOtbnEcdsaModeWords = 1,
+  /*
+   * Mode to generate a new random keypair.
+   *
+   * Value taken from `p384_ecdsa.s`.
+   */
+  kOtbnEcdsaModeKeygen = 0x3d4,
+  /*
+   * Mode to generate a signature.
+   *
+   * Value taken from `p384_ecdsa.s`.
+   */
+  kOtbnEcdsaModeSign = 0x15b,
+  /*
+   * Mode to verify a signature.
+   *
+   * Value taken from `p384_ecdsa.s`.
+   */
+  kOtbnEcdsaModeVerify = 0x727,
+};
+
+/**
+ * Makes a single dptr in the P384 library point to where its value is stored.
+ */
+static void setup_data_pointer(const otbn_addr_t dptr,
+                               const otbn_addr_t value) {
+  otbn_dmem_write(sizeof(value) / sizeof(uint32_t), &value, dptr);
+}
+
+/**
+ * Sets up all data pointers used by the P384 library to point to DMEM.
+ *
+ * The ECDSA P384 OTBN library makes use of "named" data pointers as part of
+ * its calling convention, which are exposed as symbol starting with `dptr_`.
+ * The DMEM locations these pointers refer to is not mandated by the P384
+ * calling convention; the values can be placed anywhere in OTBN DMEM.
+ *
+ * This function makes the data pointers refer to the pre-allocated DMEM
+ * regions to store the actual values.
+ */
+static void setup_data_pointers(void) {
+  setup_data_pointer(kOtbnVarEcdsaDptrX, kOtbnVarEcdsaX);
+  setup_data_pointer(kOtbnVarEcdsaDptrY, kOtbnVarEcdsaY);
+  setup_data_pointer(kOtbnVarEcdsaDptrRnd, kOtbnVarEcdsaRnd);
+  setup_data_pointer(kOtbnVarEcdsaDptrMsg, kOtbnVarEcdsaMsg);
+  setup_data_pointer(kOtbnVarEcdsaDptrR, kOtbnVarEcdsaR);
+  setup_data_pointer(kOtbnVarEcdsaDptrS, kOtbnVarEcdsaS);
+}
+
+/**
+ * Set the message digest for signature generation or verification.
+ *
+ * OTBN requires the digest in little-endian form, so this routine flips the
+ * bytes.
+ *
+ * @param digest Digest to set (big-endian).
+ * @return OK or error.
+ */
+static status_t set_message_digest(const uint32_t digest[kP384ScalarWords]) {
```

Comment:
I see that this helper function is repeated for both `sign` and `verify`; Maybe this could go in `p384_common`, and take the OTBN destination address as an argument? Alternatively, since it only has one call site in each file, it could be inlined.